Volodymyr Lut, UCU 2019

### Data overview

In this project the data from the [Ukrainian Open Data Portal](https://data.gov.ua) was used.

The dataset contains information about international bus routes and bus stops in Ukraine. Under [this link](https://data.gov.ua/dataset/d5d79cf2-ccc7-40a9-9b7f-4c5cd0d061f9) two files are allocated: `international_routes_stops3.csv` (contains information about stops and schedule of international routes) and `international_routes.csv` (contains information about starting and destination points of the routes, along with schedules). Copies of those datasets (actual on Monday, June 10 2019) are contained under the `raw_data` directory of this repository.

In [68]:
import pandas as pd

In [69]:
stops = pd.read_csv('./raw_data/international_routes_stops3.csv', sep=';', encoding='windows-1251')
stops.head()

id rejs_number stop_city_name                             stop_address  \
0  CZ1357           1           Київ  вулиця Набережно-Печерська дорога, 10-а   
1  CZ1357           1           Київ                       вул. С.Петлюри, 32   
2  CZ1357           1        Житомир                        вул. Київська, 93   
3  CZ1357           1          Рівне                        вул. Київська, 40   
4  CZ1357           1          Львів                      вул. Стрийська, 109   

    stop_code route_frequency distance_direct arrival_direct departure_direct  \
0  3200000000         Щоденно               0          20:00            20:10   
1  3200000000         Щоденно              12          20:30            21:00   
2  1810100000         Щоденно             140          23:20            23:40   
3  5610100000         Щоденно             328           2:50             3:10   
4  4610100000         Щоденно             539           6:10             6:30   

  time_diff_direct arrival_return departure_return time_diff_return  \
0                0          15:30              NaN                0   
1                0          14:40            15:00                0   
2                0          12:50            13:00                0   
3                0          10:30            10:40                0   
4                0           8:35             8:50                0   

   stop_number  Unnamed: 14  
0          0.0          NaN  
1          1.0          NaN  
2          2.0          NaN  
3          3.0          NaN  
4          4.0          NaN

In [70]:
routes = pd.read_csv('./raw_data/international_routes.csv')
routes.head()

permision_delivery_date  n/o    id  \
0                     NaN    3   MD3   
1                     NaN    8   MD8   
2              05.06.2018   11  RU11   
3                     NaN   12  RU12   
4                     NaN   13  RU13   

                                    march_route_name  \
0  АК АЕРОПОРТ БОРИСПІЛЬ (вул. Бориспільська, 7) ...   
1            Балта АС (вул. Уварова, 38а) - Дубосари   
2                                      Бахмут-Москва   
3  Бахмут (АС вул. Космонавтів,5)-Москва (Новоясе...   
4                                    Бахму - Москва    

                                        company_name  \
0                                     ФОП Гергі Є.М.   
1  ПП Фаворіт 65005, м. Одеса, вул. Бугаївська, 2...   
2                 ТОВ"Форум Авто"|ФОП"Мелкумян Ю.О."   
3                                   ТОВ "СЕВЕР-АВТО"   
4                              ФОП Базілєвський В.О.   

                               company_name_external border_crossing  \
0                            S.R.L. "SARVALTEH AUTO"          Табаки   
1  ОАО "Дубоссарское автотранспортное предприятие...       Платоново   
2                                 ООО"Газавторемонт"        Гоптівка   
3                                       без паритету      Плетенівка   
4                                  ООО "Горизон-Тур"       Танюшівка   

  border_crossing_alternative march_route_time_direct march_route_time_return  \
0                       16:00                   16:00                     NaN   
1                         NaN                     NaN                     NaN   
2                  Плетенівка                   16:50                   16:50   
3                         NaN                     NaN                     NaN   
4                         NaN                   19:30                   15:30   

  march_route_regularity        license_terms destionation_country_name  \
0                Щоденно  2020-02-08 00:00:00                   Молдова   
1                    NaN  2020-06-07 00:00:00                   Молдова   
2                Щоденно  2019-07-20 00:00:00                     Росія   
3                    NaN  2019-02-28 00:00:00                     Росія   
4                Щоденно  2019-01-07 00:00:00                     Росія   

  destionation_country_code  march_route_status  
0                        MD                True  
1                        MD                True  
2                        RU                True  
3                        RU                True  
4                        RU                True

In this project file `international_routes_stops3.csv` will be used. It should be filtered - in this research I'm interested only in routes, which are heading to Poland. 

**Explanation of useful columns:**

1. *ID* decodes the bus line id - when checking this ID in the `international_routes.csv` dataset, one could determine line operator's comoany name, information about it's license, border crossing point etc.

2. *stop_city_name* together with *stop_address* will be used to determine geolocation of the stop, which will be used for visualisation on maps.

3. *stop_number* is a column which will be used to determine the position of the stop in the route.

In [72]:
### Filtering out Polish routes
pol = stops[stops.id.str[:2].str.upper() == 'PL']
pol = pol.dropna(subset=['stop_city_name', 'stop_address'])
pol.shape

(6446, 15)

In [73]:
import numpy as np

In [74]:
np.unique(pol.id).shape

(482,)

There are **6446 stops** belonging to **482 bus lines** heading to Poland after clearing the data.

### Data Mining

For visualizations we need to know the exact location of every bus stop. In order to scrape this data, [Google Maps API](https://developers.google.com/maps) will be used.

In [75]:
pol = pol.drop("Unnamed: 14", axis=1)
pol = pol.assign(full_address=(pol.stop_city_name + ', ' + pol.stop_address).values)
pol.head()

id rejs_number       stop_city_name      stop_address   stop_code  \
2530  PL944           1               Жовква  пл. Коновальця 1  4622710100   
2531  PL944           1          Рава Руська  вул. Двірцева, 4  4622710400   
2534  PL944           1  Томашів Любельський   ul. Zamojska, 9      22-600   
2535  PL944           2               Жовква  пл. Коновальця 1  4622710100   
2536  PL944           2          Рава Руська  вул. Двірцева, 4  4622710400   

     route_frequency distance_direct arrival_direct departure_direct  \
2530         Щоденно               0           4:20             4:40   
2531         Щоденно              32           5:25             5:30   
2534         Щоденно              63           6:05              NaN   
2535         Щоденно               0          21:50            22:10   
2536         Щоденно              32          22:55            23:00   

     time_diff_direct arrival_return departure_return time_diff_return  \
2530                0          14:55              NaN                0   
2531                0          14:05            14:10                0   
2534               -1          11:00            11:30                0   
2535                0           6:25              NaN                0   
2536                0           5:35             5:40                0   

      stop_number                          full_address  
2530          0.0              Жовква, пл. Коновальця 1  
2531          1.0         Рава Руська, вул. Двірцева, 4  
2534          4.0  Томашів Любельський, ul. Zamojska, 9  
2535          5.0              Жовква, пл. Коновальця 1  
2536          6.0         Рава Руська, вул. Двірцева, 4

In [76]:
np.unique(pol.full_address).shape

(2080,)

In [77]:
import googlemaps
import re
# Uncomment line below and paste your key there
# gmaps = googlemaps.Client(key='<YOUR KEY HERE>')

In [78]:
from tqdm import tqdm

In [79]:
# Use this code to scrape geocoding data and save it to file
geocoded_addresses = []
for address in tqdm(np.unique(pol.full_address)):
    geocode_result = gmaps.geocode(address)
    geocoded_addresses.append(geocode_result)
np.savetxt("geocoded.tmp", geocoded_addresses, fmt='%s')

  0%|          | 0/2080 [00:00<?, ?it/s]


NameError: name 'gmaps' is not defined

In [80]:
import json

In [81]:
# Use this code to restore geocoding data from file
with open('geocoded.tmp') as fp:  
    geocoded_addresses = []   
    geocode_result = fp.readline()
    i = 0
    while geocode_result:
        geocode_result = geocode_result.replace("Boryspil'", "Boryspil")
        geocode_result = re.sub("(?<=[a-z])'(?=[a-z])", "", geocode_result, flags=re.IGNORECASE).replace("True", '"true"').replace("False", '"false"').replace("'", '"').replace('""', '"').replace('"-', '-')
        geocode_result = re.sub('(?<=[a-z]) "(?=[a-z])', " ", geocode_result, flags=re.IGNORECASE)
        geocode_result = re.sub('(?<=[a-z])" (?=[a-z])', " ", geocode_result, flags=re.IGNORECASE)
        geocode_result = re.sub('(?<=[a-z])", (?=[a-z])', " ", geocode_result, flags=re.IGNORECASE)
        if(i == 735):
            print(geocode_result)
        geocoded_addresses.append(json.loads(geocode_result))
        geocode_result = fp.readline()

In [112]:
full_addresses = pol.full_address
unique_addresses = np.unique(full_addresses).tolist()
geolocation_list = []
for address in tqdm(full_addresses):
    i = unique_addresses.index(address)
    try :
        lat = geocoded_addresses[i][0]['geometry']['location']['lat']
        lng = geocoded_addresses[i][0]['geometry']['location']['lng']
        country = geocoded_addresses[i][0]['address_components'][-2]['long_name']
        administrative_level_1 = geocoded_addresses[i][0]['address_components'][-3]['long_name']
        administrative_level_2 = geocoded_addresses[i][0]['address_components'][-4]['long_name']
        geolocation_list.append([
            lat,
            lng,
            country,
            administrative_level_1,
            administrative_level_2
        ])
    except IndexError :
        geolocation_list.append([None, None, None, None])
        print (address, i)

geocoded_dataframe = pd.DataFrame(geolocation_list, columns = ['lat', 'lng', 'country', 'administrative_level_1', 'administrative_level_2'])

100%|██████████| 6446/6446 [00:00<00:00, 30577.27it/s]

Ельбонг, ul. Grunwaldzka, 14 490
Минськ Мазовецький, pl.Dworcowa,1 1208
Замостя, вул. Грубешівська, 1 557
Червінськ-над-Віслою, вул. В. Ягели, 18 1819
Монастирська, вул. Шевченка, 57 1212
Пила, ul. Kvitova, 18 1376
м. Грудзьонз, ul. Dworcowa, 1 1950
Нови-Двур-Мазовецьки "Модлин", ul.gen.Wiktora Thommee 1241
Нови-Двур-Мазовецьки "Модлин", ul.gen.Wiktora Thommee 1241
Петрикув Трибунальський, вул.Польської Організації Військової 1370
Монастирська, вул. Шевченка, 57 1212
м. Грудзьонз, ul. Dworcowa, 1 1950
Нови-Двур-Мазовецьки "Модлин", ul.gen.Wiktora Thommee 1241
Бердичів, пл. Привокзальна, 1-А 49
вул.Звірки і Вігури,1, Dw. PKS "Zachodnia", str. Al. Jerozlimskie,144 1925
Грудзонз, ul. Rapatskogo 33 428
Нови-Двур-Мазовецьки "Модлин", ul.gen.Wiktora Thommee 1241
Нови-Двур-Мазовецьки "Модлин", ul.gen.Wiktora Thommee 1241
Грудзонз, ul. Rapatskogo 33 428
Ельбонг, plac Dworcowy 489
Нови-Двур-Мазовецьки "Модлин", ul.gen.Wiktora Thommee 1241
Грудзонз, Dworcowa,42 427
Грудзенз, Dworcowa,42 425
Крот

(6446, 5)

In [121]:
poland_geocoded = pd.concat([pol.reset_index(drop=True), geocoded_dataframe.reset_index(drop=True)], axis=1)

In [124]:
poland_geocoded.head()

id rejs_number       stop_city_name      stop_address   stop_code  \
0  PL944           1               Жовква  пл. Коновальця 1  4622710100   
1  PL944           1          Рава Руська  вул. Двірцева, 4  4622710400   
2  PL944           1  Томашів Любельський   ul. Zamojska, 9      22-600   
3  PL944           2               Жовква  пл. Коновальця 1  4622710100   
4  PL944           2          Рава Руська  вул. Двірцева, 4  4622710400   

  route_frequency distance_direct arrival_direct departure_direct  \
0         Щоденно               0           4:20             4:40   
1         Щоденно              32           5:25             5:30   
2         Щоденно              63           6:05              NaN   
3         Щоденно               0          21:50            22:10   
4         Щоденно              32          22:55            23:00   

  time_diff_direct arrival_return departure_return time_diff_return  \
0                0          14:55              NaN                0   
1                0          14:05            14:10                0   
2               -1          11:00            11:30                0   
3                0           6:25              NaN                0   
4                0           5:35             5:40                0   

   stop_number                          full_address        lat        lng  \
0          0.0              Жовква, пл. Коновальця 1  50.056643  23.972478   
1          1.0         Рава Руська, вул. Двірцева, 4  50.230070  23.636844   
2          4.0  Томашів Любельський, ul. Zamojska, 9  50.454665  23.419679   
3          5.0              Жовква, пл. Коновальця 1  50.056643  23.972478   
4          6.0         Рава Руська, вул. Двірцева, 4  50.230070  23.636844   

   country          region                  city  
0  Ukraine  Lvivska oblast  Zhovkivskyi district  
1  Ukraine  Lvivska oblast  Zhovkivskyi district  
2   Poland       lubelskie           tomaszowski  
3  Ukraine  Lvivska oblast  Zhovkivskyi district  
4  Ukraine  Lvivska oblast  Zhovkivskyi district

In [118]:
poland_geocoded.to_csv('poland_bus_routes_geocoded.csv')

lat        lng  country                           region  \
0   50.056643  23.972478  Ukraine                   Lvivska oblast   
1   50.230070  23.636844  Ukraine                   Lvivska oblast   
2   50.454665  23.419679   Poland                        lubelskie   
3   50.056643  23.972478  Ukraine                   Lvivska oblast   
4   50.230070  23.636844  Ukraine                   Lvivska oblast   
5   50.454665  23.419679   Poland                        lubelskie   
6   50.056643  23.972478  Ukraine                   Lvivska oblast   
7   50.230070  23.636844  Ukraine                   Lvivska oblast   
8   50.454665  23.419679   Poland                        lubelskie   
9   50.056643  23.972478  Ukraine                   Lvivska oblast   
10  50.230070  23.636844  Ukraine                   Lvivska oblast   
11  50.454665  23.419679   Poland                        lubelskie   
12  50.442224  30.491251  Ukraine                        Kyiv City   
13  50.268670  28.692149  Ukraine               Zhytomyrska oblast   
14  49.236680  28.482954  Ukraine                 Vinnytska oblast   
15  49.237034  28.403936  Ukraine                 Vinnytska oblast   
16  49.381850  27.614116  Ukraine               Khmelnytska oblast   
17  49.427435  27.033709  Ukraine               Khmelnytska oblast   
18  49.536766  26.222091  Ukraine               Khmelnytska oblast   
19  49.526965  26.120095  Ukraine               Ternopilska oblast   
20  49.543930  25.596769  Ukraine               Ternopilska oblast   
21  49.808147  24.905977  Ukraine                   Lvivska oblast   
22  49.839932  23.993766  Ukraine                   Lvivska oblast   
23  50.042241  22.004784   Poland         Województwo podkarpackie   
24  50.067237  19.949116   Poland                      małopolskie   
25  50.261637  19.019301   Poland                          śląskie   
26  50.806335  19.118512   Poland                          śląskie   
27  51.074524  19.437829   Poland              województwo łódzkie   
28  51.410527  19.685797   Poland              województwo łódzkie   
29  51.759300  19.431600   Poland              województwo łódzkie   
..        ...        ...      ...                              ...   
70  51.246454  22.568446   Poland               Lublin Voivodeship   
71  52.218443  20.963996   Poland                      mazowieckie   
72  53.013790  18.598444   Poland  Kuyavian-Pomeranian Voivodeship   
73  53.123480  18.008438   Poland  Kuyavian-Pomeranian Voivodeship   
74  54.352025  18.646638   Poland           Pomeranian Voivodeship   
75  54.518890  18.530541   Poland           Pomeranian Voivodeship   
76  49.839020  23.994397  Ukraine                   Lvivska oblast   
77  54.464148  17.028482   Poland           Pomeranian Voivodeship   
78  54.194322  16.171491   Poland      West Pomeranian Voivodeship   
79  53.428544  14.552812   Poland      West Pomeranian Voivodeship   
80  48.274273  25.990223  Ukraine              Chernivetska oblast   
81  48.264913  25.951979  Ukraine              Chernivetska oblast   
82  48.436150  25.761404  Ukraine              Chernivetska oblast   
83  48.641623  25.735146  Ukraine               Ternopilska oblast   
84  49.005525  25.801987  Ukraine               Ternopilska oblast   
85  49.302116  25.689442  Ukraine               Ternopilska oblast   
86  49.543930  25.596769  Ukraine               Ternopilska oblast   
87  49.786820  24.016383  Ukraine                   Lvivska oblast   
88  49.839932  23.993766  Ukraine                   Lvivska oblast   
89  50.042775  22.006881   Poland         Województwo podkarpackie   
90  50.067237  19.949116   Poland                      małopolskie   
91  50.262740  19.017119   Poland                          śląskie   
92  50.662778  17.929210   Poland                         opolskie   
93  51.096292  17.037621   Poland         Województwo dolnośląskie   
94  49.839050  24.030761  Ukraine                   Lvivska oblast   
95  49.442814  23.00

In [117]:
full_addresses

2530                              Жовква, пл. Коновальця 1
2531                         Рава Руська, вул. Двірцева, 4
2534                  Томашів Любельський, ul. Zamojska, 9
2535                              Жовква, пл. Коновальця 1
2536                         Рава Руська, вул. Двірцева, 4
2539                  Томашів Любельський, ul. Zamojska, 9
2540                              Жовква, пл. Коновальця 1
2541                         Рава Руська, вул. Двірцева, 4
2544                  Томашів Любельський, ul. Zamojska, 9
2545                              Жовква, пл. Коновальця 1
2546                         Рава Руська, вул. Двірцева, 4
2549                  Томашів Любельський, ul. Zamojska, 9
2550                             Київ, вул. С. Петлюри, 32
2551                            Житомир, вул. Київська, 93
2552                            Вінниця, вул. Киїівська, 8
2553                        Вінниця, Хмельницьке шосе, 107
2554                         Летичів, вул. Ю.Савицького,